### PR TITLE
Make HostDBInfo class safer to use.

### PR DESCRIPTION
**Pull request by ywkaras**

### CODE REVIEW

**Comment by ywkaras** [comment]
I don't see any current potential memory corruption with this class, but I think this PR could help prevent potential future mistakes.

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by maskit** [comment]
[approve ci autest]

Rerunning autest because the last run didn't check anything.
```
fatal: ambiguous argument '3008d8c7cc1234394d9378453dca30f4d1803455^...3008d8c7cc1234394d9378453dca30f4d1803455': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
No relevant files changed, skipping run
Setting status of 3008d8c7cc1234394d9378453dca30f4d1803455 to SUCCESS with url https://ci.trafficserver.apache.org/job/autest-github/13414/ and message: 'Build finished. '
Using context: Jenkins autest
Finished: SUCCESS
```

File: iocore/hostdb/I_HostDBProcessor.h [hunk 1]
```diff
@@ -146,16 +146,17 @@ struct HostDBInfo : public RefCountObj {
...
+    ret->_iobuffer_index = iobuffer_index;
     return ret;
   }
 
   void
   free() override
   {
-    Debug("hostdb", "freeing %d bytes at [%p]", (1 << (7 + iobuffer_index)), this);
-    ioBufAllocator[iobuffer_index].free_void((void *)(this));
+    ink_release_assert(from_alloc());
```
Comment (zizhong) [inline iocore/hostdb/I_HostDBProcessor.h:157]: Looks like, the diff introduces new ink_release_assert to make sure the object is allocated from HostDBInfo::alloc().
Could we simply the object uncopyable to achieve the same?

Comment (zizhong) [reply iocore/hostdb/I_HostDBProcessor.h:157]: >The current code requires that instances of HostDBInfo be copyable.
Could you elaborate on that? I'll be surprised to see if any HostDBInfo is copied around. ^_^

Comment (ywkaras) [reply iocore/hostdb/I_HostDBProcessor.h:157]: Here it the case I am aware of where a copy occurs:  https://github.com/apache/trafficserver/blob/af9c4f39aeb40515bb42469cb0aeeb0ac27fec54/proxy/http/HttpSM.cc#L2254

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Pull request merged**